components/ephemeral_storage/browser/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,88 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/ephemeral_storage/browser/ephemeral_storage_tab_helper.h"
+
+#include <string>
+#include "base/feature_list.h"
+#include "components/content_settings/core/common/features.h"
+#include "content/browser/dom_storage/dom_storage_context_wrapper.h"
+#include "content/browser/renderer_host/render_view_host_delegate.h"
+#include "content/browser/renderer_host/render_view_host_impl.h"
+#include "content/browser/storage_partition_impl_map.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/render_view_host.h"
+#include "content/public/browser/web_contents.h"
+
+using content::BrowserContext;
+using content::DOMStorageContextWrapper;
+using content::NavigationHandle;
+using content::SiteInstance;
+using content::StoragePartition;
+using content::WebContents;
+
+namespace ephemeral_storage {
+
+EphemeralStorageTabHelper::~EphemeralStorageTabHelper() {}
+
+EphemeralStorageTabHelper::EphemeralStorageTabHelper(WebContents* web_contents)
+    : WebContentsObserver(web_contents) {}
+
+void EphemeralStorageTabHelper::ReadyToCommitNavigation(
+    NavigationHandle* navigation_handle) {
+  if (!navigation_handle->IsInMainFrame())
+    return;
+  if (navigation_handle->IsSameDocument())
+    return;
+
+  ClearEphemeralStorage();
+}
+
+void EphemeralStorageTabHelper::WebContentsDestroyed() {
+  ClearEphemeralStorage();
+}
+
+void EphemeralStorageTabHelper::ClearEphemeralStorage() {
+  if (!base::FeatureList::IsEnabled(content_settings::kBraveEphemeralStorage)) {
+    return;
+  }
+
+  WebContents* contents = web_contents();
+  BrowserContext* browser_context = contents->GetBrowserContext();
+  SiteInstance* site_instance = contents->GetSiteInstance();
+  StoragePartition* storage_partition =
+      BrowserContext::GetStoragePartition(browser_context, site_instance);
+  DOMStorageContextWrapper* dom_storage_context =
+      static_cast<DOMStorageContextWrapper*>(
+          storage_partition->GetDOMStorageContext());
+
+  storage::mojom::SessionStorageControl* session_storage_control =
+      dom_storage_context->GetSessionStorageControl();
+  if (!session_storage_control)
+    return;
+
+  std::string session_namespace_id =
+      contents->GetRenderViewHost()
+          ->GetDelegate()
+          ->GetSessionStorageNamespace(site_instance)
+          ->id() +
+      "ephemeral-session-storage";
+  session_storage_control->DeleteNamespace(session_namespace_id,
```

Comment:
I think you should be using DOMStorageContext::DeleteSessionStorage/DeleteLocalStorage to avoid using anything in `content/browser` which isn't allowed as a dep here

Reply:
`DOMStorageContext::DeleteLocalStorage` takes an origin and we need to delete the `LocalStorage` for every third-party origin. I'm not sure it's appropriate here.

Reply:
In the latest version of the change, I've moved this to `BrowserContext` to avoid the layering violation.